### PR TITLE
Add tests for pin linting, fix issue #190.

### PR DESCRIPTION
**Pull request by johanneskoester**

### CODE REVIEW

File: bioconda_utils/lint_functions.py [hunk 1]
```diff
@@ -288,17 +288,30 @@ def _pin(env_var, dep_name):
...
+        in_requirements = False
+        in_run = False
         for line in open(os.path.join(recipe, 'meta.yaml')):
+            if line.startswith("requirements:"):
+                in_requirements = True
+            elif in_requirements and line.strip().startswith("run:"):
+                in_run = True
+            elif in_requirements and line.strip().startswith("build:"):
+                in_run = False
+            elif not line.startswith(" "):
```
Comment (mbargull) [inline bioconda_utils/lint_functions.py]: Should/could be `elif not line.startswith((" ", "#")):`, right? That way, not indented comments are also handled.

Comment (johanneskoester) [reply bioconda_utils/lint_functions.py]: Good point.

File: bioconda_utils/lint_functions.py [hunk 2]
```diff
@@ -288,17 +288,30 @@ def _pin(env_var, dep_name):
...
+            elif in_requirements and line.strip().startswith("build:"):
+                in_run = False
+            elif not line.startswith(" ") and not line.startswith("#"):
+                in_requirements = False
+                in_run = False
             line = line.strip()
-            if line.startswith('- {}'.format(dep_name)):
-                if env_var not in line: # and '{{' not in line and '}}' not in line:
+            if in_run and line.startswith('- {}'.format(dep_name)):
+                if pin_pattern.search(line):
```
Comment (nsoranzo) [inline bioconda_utils/lint_functions.py]: Shouldn't this line end with ```is None:``` ?

Comment (johanneskoester) [reply bioconda_utils/lint_functions.py]: Yes it should! Thanks!

Comment (johanneskoester) [reply bioconda_utils/lint_functions.py]: That also explains the failing tests.

**Comment by nsoranzo** [comment]
Should the linting be applied to both `run` and `build` requirements or just `run`? Because the tests seem to specify the former while the code implements the latter.

**Comment by johanneskoester** [comment]
In fact it is a bit more complicated. If the dependency does not occur in run, we don't need to pin it at all because it is statically linked. However, my previous implementation did not explicitly check all cases, which I have fixed in the latest commit.

File: bioconda_utils/lint_functions.py [hunk 3]
```diff
@@ -333,7 +357,5 @@ def pin(recipe, meta, df):
     _pin('CONDA_HDF5', 'hdf5'),
     _pin('CONDA_NCURSES', 'ncurses'),
     _pin('CONDA_HTSLIB', 'htslib'),
-
-    # Pinning in env_matrix.yml is not consistent for bzip2.
-    # _pin('CONDA_BZIP2', 'bzip2'),
+    _pin('CONDA_BZIP2', 'bzip2'),
```
Comment (nsoranzo) [inline bioconda_utils/lint_functions.py]: I think that need to be special-cased in the code above. Maybe also a specific test should be added.

Comment (johanneskoester) [reply bioconda_utils/lint_functions.py]: Why? The wrong entry in the env matrix was fixed in PR bioconda/bioconda-recipes#6310

Comment (nsoranzo) [reply bioconda_utils/lint_functions.py]: Ah, sorry, I missed that PR!

**Pull request merged**